python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+    """Configuration class for generating integer IDs from
+    ``{visit, detector}`` data IDs.
+
+    See `IdGenerator` for usage.
+    """
+
+    packer = Instrument.make_dimension_packer_config_field()
+
+    def _make_dimension_packer(self, data_id: DataCoordinate) -> DimensionPacker:
+        # Docstring inherited.
+        return self.packer.apply(data_id, is_exposure=False)
+
+
+class SkyMapIdGeneratorConfig(BaseIdGeneratorConfig):
+    """Configuration class for generating integer IDs from
+    ``{tract, patch, [band]}`` data IDs.
+
+    See `IdGenerator` for usage.
+    """
+
+    packer = SkyMapDimensionPacker.make_config_field()
+
+    def _make_dimension_packer(self, data_id: DataCoordinate) -> DimensionPacker:
+        # Docstring inherited.
+        return self.packer.apply(data_id)
+
+
+class IdGenerator:
+    """A helper class for packing some combination of a data ID, a per-data-ID
+    counter, and a release ID into a single 64-bit integer.
+
+    Notes
+    -----
+    Instances of this class are expected to usually be created via
+    configuration, which will return a derived instance.  This pattern starts
+    with one of `ExposureIdGeneratorConfig`, `VisitIdGeneratorConfig`, and
+    `SkyMapIdGeneratorConfig` (which have the same interface), and looks
+    something this:
+
+        from lsst.afw.image import SourceTable, SourceCatalog
+        from lsst.meas.base import VisitIdGeneratorConfig
+        from lsst.obs.base import Instrument
+        from lsst.pex.config import Config from lsst.pipe.base import Task
+
+        class SomeTaskConfig(Config):
+            id_generator = VisitIdGeneratorConfig.make_field()
+
+        class SomeTask(Task):
+
+            ConfigClass = SomeTaskConfig
+
+            ...
+
+            def run(self, ..., data_id: DataCoordinate):
+                id_generator = self.config.apply(data_id)
+                catalog = id_generator.make_source_catalog(self.schema) ...
+
+    Code that wishes to instead unpack these record IDs to obtain the release
+    ID, data ID and counter value should use the same config (often loaded from
+    the ``Butler``) and pass a fully-expanded data ID identifying only a
+    particular ``skymap`` or ``instrument`` to `unpacker_from_config`::
+
+        config = butler.get("some_task_config")
+        catalog = butler.get("some_output_catalog", given_data_id)
+        unpacker = IdGenerator.unpacker_from_config(
+            config.id_generator, butler.registry.expandDataId(skymap="HSC"),
+        )
+        release_id, embedded_data_id, counter = unpacker(catalog[0]["id"])
+        assert embedded_data_id == given_data_id
+
+    This example is a bit contrived, as the ability to reconstruct the data ID
+    is really only useful when you don't have it already, such as when the
+    record ID is obtained from some further-processed version of the original
+    table (such as a SQL database), and in that context the right config to
```

Comment:
The example includes a butler because I figured it was the most common way to get the right config, but you don't _need_ a butler to do the unpacking.  Just this morning @natelust was in my office talking about an idea to make it possible to get tasks (and by extensions, configs) by reading pipeline files, which would be another nice code path.

Neither of those is an ideal path for science users, at least not without some convenience code layered on top (some e.g.  `unpack_object_id` function that knows what config dataset types or pipelines to go read).  But I think the right answer for science users will have to come down to some sort of convenience code.